### PR TITLE
simplify the train of sq

### PR DESCRIPTION
**Pull request by LHT129**
- all kinds of sq have similar train process
- add extend for special sq trainer

### CODE REVIEW

File: src/quantization/scalar_quantization_trainer.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    }
+}
+
+uint64_t
+ScalarQuantizationTrainer::sample_train_data(const float* data,
+                                             uint64_t count,
+                                             std::vector<float>& sample_datas,
+                                             bool need_normalize) {
+    uint64_t step = 2147483647UL % count;
+    auto ret = max_sample_count_;
```
Comment (jiaweizone) [inline src/quantization/scalar_quantization_trainer.cpp]: ret -> sample_count ?

File: src/quantization/scalar_quantization_trainer.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ScalarQuantizationTrainer::classic_train(const float* data,
+                                         uint64_t count,
+                                         float* upper_bound,
+                                         float* lower_bound) {
+    for (uint64_t i = 0; i < dim_; ++i) {
+        upper_bound[i] = std::numeric_limits<float>::lowest();
+        lower_bound[i] = std::numeric_limits<float>::max();
+        for (uint64_t j = 0; j < count; ++j) {
+            auto value = data[j * dim_ + i];
+            upper_bound[i] = std::max(upper_bound[i], value);
```
Comment (jiaweizone) [inline src/quantization/scalar_quantization_trainer.cpp:71]: minor: Here change to check and update, is it possible to reduce unnecessary memory write ?

File: src/quantization/scalar_quantization_trainer.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                                        uint64_t count,
+                                        float& upper_bound,
+                                        float& lower_bound,
+                                        bool need_normalize,
+                                        ScalarQuantizationTrainer::SQTrainMode mode) {
+    std::vector<float> sample_datas;
+    auto sample_count = this->sample_train_data(data, count, sample_datas, need_normalize);
+    std::vector<float> upper(dim_);
+    std::vector<float> lower(dim_);
+    if (mode == CLASSIC) {
```
Comment (jiaweizone) [inline src/quantization/scalar_quantization_trainer.cpp:54]: ditto

File: src/quantization/scalar_quantization_trainer.cpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+void
+ScalarQuantizationTrainer::Train(const float* data,
+                                 uint64_t count,
+                                 float* upper_bound,
+                                 float* lower_bound,
+                                 bool need_normalize,
+                                 ScalarQuantizationTrainer::SQTrainMode mode) {
+    std::vector<float> sample_datas;
+    auto sample_count = this->sample_train_data(data, count, sample_datas, need_normalize);
+    if (mode == CLASSIC) {
```
Comment (jiaweizone) [inline src/quantization/scalar_quantization_trainer.cpp:38]: add unsupported mode check and print warn log ?

File: src/quantization/scalar_quantization_trainer.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        this->classic_train(sample_datas.data(), sample_count, upper_bound, lower_bound);
+    }
+}
+
+void
+ScalarQuantizationTrainer::TrainUniform(const float* data,
+                                        uint64_t count,
+                                        float& upper_bound,
+                                        float& lower_bound,
```
Comment (jiaweizone) [inline src/quantization/scalar_quantization_trainer.cpp:47]: upper_bound / lower_bound change to `float*`

File: src/quantization/scalar_quantization_trainer.cpp (hunk 1 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp]: done

File: src/quantization/scalar_quantization_trainer.cpp (hunk 2 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp:71]: I think compiler may optimize it, the upper_bound[i] may in register

File: src/quantization/scalar_quantization_trainer.cpp (hunk 5 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp:47]: must be value not a array, so pointer may not good

File: src/quantization/scalar_quantization_trainer.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    int dim_{0};
+
+    int bits_{8};
+
+    uint64_t max_sample_count_{MAX_DEFAULT_SAMPLE};
+
+    const static uint64_t MAX_DEFAULT_SAMPLE{65536};
+};
+
+}  // namespace vsag
```
Comment (ShawnShawnYou) [inline src/quantization/scalar_quantization_trainer.h]: add one line

File: src/quantization/scalar_quantization_trainer.cpp (hunk 4 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.cpp:38]: this is inner mode, I just want to extend it in future

File: src/quantization/scalar_quantization_trainer.h (hunk 6 above)
Comment (LHT129) [reply src/quantization/scalar_quantization_trainer.h]: done

**Review by ShawnShawnYou — APPROVED** [review]

**Review by wxyucs — APPROVED** [review]
lgtm

File: src/quantization/scalar_quantization_trainer.cpp (hunk 4 above)
Comment (jiaweizone) [reply src/quantization/scalar_quantization_trainer.cpp:38]: Ok, remember to add check

File: src/quantization/scalar_quantization_trainer.cpp (hunk 5 above)
Comment (jiaweizone) [reply src/quantization/scalar_quantization_trainer.cpp:47]: Ok, please move the output params `upper_bound / lower_bound` to params list end

**Review by jiaweizone — APPROVED** [review]

**Pull request merged**